### ISSUE TITLE
Migrate to JSON internal schema.

### ISSUE THREAD

**Issue by subdavis** [VIAME/VIAME-Web#96]
It's wasteful and slow to convert CSV to JSON every time the annotation page is loaded.  We probably shouldn't read bytes from disk at all on the server, and instead let the client directly request a girder file.

IMO, the internal representation of tracks should be some JSON schema.  It would be simple enough to do a copy-on-read the first time the client requests a VIAME csv, then maintain the JSON data for annotation, and add export endpints to convert back to CSV if the user wants it.

This would also allow us to support other necessary annotation data types, like `kw18` and `kpf`, which are needed for general adoption within kitware.

**Comment by AlmightyYakob** [comment]
When it comes to training, there's currently no supported JSON format for annotation in the training step of viame (it [supports](https://github.com/VIAME/VIAME/tree/master/examples/detection_file_conversions#integrated-detection-formats) csv, kw18, and HABCAM). So we'd definitely need to have the conversion in place for training.

**Comment by AlmightyYakob** [comment]
I should add though I'm definitely in favor of us doing this.

**Comment by subdavis** [comment]
I fully expect that any interaction with external systems would involve conversion in and out of our internal schema.

**Comment by mattdawkins** [comment]
Training now supports json, the main issue is it needs to auto-detect json vs csv as csv is still in production (and probably always will be as we have some contracts which require it for excel users doing analytics on the output).

No objection to changing what is used internally as long as we always have optional support for csv at input (initial upload) and output (download through GUI) time as well.

**Comment by subdavis** [comment]
Semi-related:

Ive been thinking a lot about how detections and tracks are represented in the client, and I think we may want to migrate to a track-centric data structure.

* representing tracks as a global detection array cannot ever support interpolation
* a flat detection array is efficient O(1) for frame lookup, but bad for almost everything else.
* For example, update/insert is always O(N) (where N is detection count)
* This O(N) cost is multiplied in practice because there are several different "views" of the data needed that all get computed on update: like the unfiltered track list for the left pane, the filtered detection list for rendering, etc.
* Right now the lookup cost for figuring out a track's start and end frames is also O(N) which is why  gotoTrackFirstFrame is such a pain:

```
function gotoTrackFirstFrame({ trackId }) {
      setTrackEditMode(trackId, false);
      const _frame = eventChartData.value.find((d) => d.track === trackId)
        .range[0];
      seek(_frame);
    }
```

This function uses `eventChartData`, which has an update cost of O(N), and isn't correct to use here because it doesn't have info about tracks that were filtered out due to low confidence.

It is necessary to have a list of tracks that looks something like this:

```
{
  'min_confidence': Number,
  'max_confidence': Number,
  'begin': Frame,
  'end': Frame,
  'detections': Array<Detection>,
}
```

This is an issue because the lookup cost, depending on the track/detection ratio, which is 1:1 for pipelines where only detections were run, can be up to O(N) instead of O(1).

Changing the structure like this would flip the update/insert cost to something roughly constant, and lookup cost could be made roughly logarithmic if we had something like an interval tree: https://www.cs.cmu.edu/~ckingsf/bioinfo-lectures/intervaltrees.pdf

It would be easy enough to construct a simple tree, but a self-balancing tree would be better, like a balanced interval tree or a red-black interval tree. 

Having previously implemented the idiot-version (iterate over tracks, filtering by range), I know that worst-case performance is actually already pretty good, so I feel pretty confident that the tree would perform well.

**Comment by subdavis** [comment]
closed in #137
